portable/NetworkInterface/TM4C/NetworkInterface.c

```diff
@@ -0,0 +1,599 @@
+
+    for (i = 0; i < SYSCONFIG_FREERTOS_TCP_RX_DESC_COUNT; i++)
+    {
+        stack_descriptor = pxGetNetworkBufferWithDescriptor(ipTOTAL_ETHERNET_FRAME_SIZE, 0);
+
+        configASSERT(NULL != stack_descriptor);
+
+        // Get a buffer from the stack and assign it to the DMA Descriptor
+        _rx_descriptors[i].pvBuffer1 = stack_descriptor->pucEthernetBuffer;
+
+        // Inform the DMA controller that the descriptors are chained and the size of the buffer
+        _rx_descriptors[i].ui32Count = DES1_RX_CTRL_CHAINED | ((buffer_size_requested << DES1_TX_CTRL_BUFF1_SIZE_S) & DES1_TX_CTRL_BUFF1_SIZE_M);
+
+        // Give the DMA descriptor to the DMA controller
+        _rx_descriptors[i].ui32CtrlStatus = DES0_RX_CTRL_OWN;
+
+        // Set the next link the DMA descriptor chain
+        _rx_descriptors[i].DES3.pLink = ((i == (SYSCONFIG_FREERTOS_TCP_RX_DESC_COUNT - 1)) ? &_rx_descriptors[0] : &_rx_descriptors[i + 1]);
+    }
+
+    // Set the RX descriptor index
+    _rx_descriptor_list.write = 0;
+
+    // Set the head of the DMA descriptor list in the EMAC peripheral
+    MAP_EMACTxDMADescriptorListSet(EMAC0_BASE, &_tx_descriptors[0]);
+    MAP_EMACRxDMADescriptorListSet(EMAC0_BASE, &_rx_descriptors[0]);
+}
+
+void freertos_tcp_ethernet_int(void)
+{
+    uint32_t status;
+
+    // Read the interrupt status
+    status = EMACIntStatus(EMAC0_BASE, true);
+
+    // Handle power management interrupts
+    if(status & EMAC_INT_POWER_MGMNT)
+    {
+        MAP_EMACTxEnable(EMAC0_BASE);
+        MAP_EMACRxEnable(EMAC0_BASE);
+
+        MAP_EMACPowerManagementStatusGet(EMAC0_BASE);
+
+        status &= ~(EMAC_INT_POWER_MGMNT);
+    }
+
+    if (status)
+    {
+        MAP_EMACIntClear(EMAC0_BASE, status);
+    }
+
+    // Handle PHY interrupts
+    if (EMAC_INT_PHY & status)
+    {
+        _process_phy_interrupt();
+    }
+
+    // Handle Transmit Complete interrupts
+    if (EMAC_INT_TRANSMIT & status)
+    {
+        _process_transmit_complete();
+    }
+
+    // Handle Receive interrupts
+    if ((EMAC_INT_RECEIVE | EMAC_INT_RX_NO_BUFFER | EMAC_INT_RX_STOPPED) & status)
+    {
+        _process_received_packet();
+    }
+}
+
+static void _process_transmit_complete(void)
+{
+    uint32_t i;
+    tEMACDMADescriptor *dma_descriptor;
+    NetworkBufferDescriptor_t *stack_descriptor;
+
+    for (i = 0; ((i < _tx_descriptor_list.number_descriptors) && (_tx_descriptor_list.read != _tx_descriptor_list.write)); i++)
+    {
+        // Get a reference to the current DMA descriptor
+        dma_descriptor = &_tx_descriptors[_tx_descriptor_list.read];
+
+        // If the descriptor is still owned by the DMA controller, exit
+        if (dma_descriptor->ui32CtrlStatus & DES0_TX_CTRL_OWN)
+        {
+            break;
+        }
+
+        // Get the 'hidden' reference to the stack descriptor from the buffer
+        stack_descriptor = *((NetworkBufferDescriptor_t **) (((uint8_t *) dma_descriptor->pvBuffer1) - ipBUFFER_PADDING));
```

Comment:
A beautiful line of code, but we have a standard function that does the same, along with some security checks.
It is this function:
~~~c
    NetworkBufferDescriptor_t * pxPacketBuffer_to_NetworkBuffer( const void * pvBuffer );

    /* Find the Network packet that contains the buffer: */
    stack_descriptor = pxPacketBuffer_to_NetworkBuffer( dma_descriptor->pvBuffer1 );
~~~
You could even add a `configASSERT(stack_descriptor!=NULL)`, because there is something really wrong if this fails.

Reply:
That would make things much easier! The porting guide for FreeRTOS+TCP does not mention this, and I'll be honest in that I hadn't looked at all the other implementations. I will make this change as well.